File: Java/BST.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package src;
/**
 *
 * @author thwilliams
 */
public class BST {
    private Node root;
    private int nodeCount;
    public BST(int r){
        Node root=new Node(r);
        this.root=root;
    }
    public BST(){
        this.root=null;
    }
    public void inOrderTraverseTree(){
        inOrderTraverseTree(this.root);
    }
    public void inOrderTraverseTree(Node focusNode) {
	        if (focusNode != null) {
	            // Traverse the left node
	            inOrderTraverseTree(focusNode.getLeft());
	            // Visit the currently focused on node
	            System.out.println(focusNode);
	            // Traverse the right node
	            inOrderTraverseTree(focusNode.getRight());
	        }
	    }
    public void preorderTraverseTree(){
         preorderTraverseTree(this.root);
    }
    public void preorderTraverseTree(Node focusNode){
	        if (focusNode != null) {
	            System.out.println(focusNode);
	            preorderTraverseTree(focusNode.getLeft());
	            preorderTraverseTree(focusNode.getRight());
	        }
	    }
    public void postOrderTraverseTree(){
        postOrderTraverseTree(this.root);
    }
    public void postOrderTraverseTree(Node focusNode) {
	        if (focusNode != null) {
	            postOrderTraverseTree(focusNode.getLeft());
	            postOrderTraverseTree(focusNode.getRight());
	            System.out.println(focusNode);
	        }
	    }
    public boolean contains(int n){
        Queue a=new Queue(this.root);
        Node k;
        while(!a.NisEmpty()){
            k=a.Ndequeue();
           if(k!=null){
               if(k.value()==n){
                   return true;
               } 
            }
            if(k.getLeft()!=null){
                a.enqueue(k.getLeft());
            }
            if(k.getRight()!=null){
                a.enqueue(k.getRight());
            }
        }
        return false;
    }
    public void addNode(int n){
        Node a=new Node(n);
        if(this.root==null){
            this.root=a;
        }else{
            Node curr=this.root;
            //to hold parent later
            Node parent;
            while(true){
                parent=curr;
                if(n<curr.value()){
                    curr=curr.getLeft();
                    if(curr==null){
                        parent.setLeft(a);
                        return;
                    }
                }else{
                    curr=curr.getRight();
                    if(curr==null){
                        parent.setRight(a);
                        return;
                    }
                }
            }
        }
    }
    public boolean remove(int d){
        Node curr=this.root;
        Node parent=this.root;
        boolean isLeft=true;
        while(curr.value()!=d){
            parent=curr;
            if(d<curr.value()){
                isLeft=true;
                curr=curr.getLeft();
                //go left
            }else{
                isLeft=false;
                curr=curr.getRight();
                //go right
            }
            if(curr==null){
            //element not inside the tree
            return false;
            }
        }  
        //no left and no right child
        if(curr.getLeft()==null && curr.getRight()==null){
            if(curr.value()==this.root.value()){
                this.root=null;
            }else if(isLeft){
                //if the node we are removing is left remove it
                parent.setLeft(null);
            }else{
                //if the node we are removing is right remove it
                parent.setRight(null);
            }
            //only is a left child
        }else if(curr.getRight()==null){
            //no right child
            //if we are removing root then get the left child
            if(curr.value()==this.root.value()){
               this.root=this.root.getLeft();
            }else if(isLeft){
                parent.setLeft(curr.getLeft());
            }else{
                parent.setRight(curr.getRight());
            }
        }else if(curr.getLeft()==null){
             //no left child, only a right child
            //if we are removing root then get the right child
            if(curr.value()==this.root.value()){
               this.root=curr.getRight();
            }else if(isLeft){
                parent.setLeft(curr.getRight());
            }else{
                parent.setRight(curr.getLeft());
            }
        }else{
             //there is both a left or right node
             Node replacement=getReplacement(curr);
              if(curr.value()==this.root.value()){
                    this.root=replacement;
             }else if(isLeft){
                    parent.setLeft(replacement);
            }else{
                     parent.setRight(replacement);
            }
            replacement.setLeft(curr.getLeft());
        }
        return true;
    }
    public Node getReplacement(Node a){
        Node parent=a;
        Node replacement=a;
        //always replace with right
        Node curr=a.getRight();
        while(curr!=null){
            parent=replacement;
            replacement=curr;
            curr=curr.getLeft();
        }
        if(replacement.value()!=a.getRight().value()){
            parent.setLeft(replacement.getRight());
            replacement.setRight(a.getRight());
        }
        return replacement;
    }
   public Node leftmost(Node a){
        Node curr=a;
        while(curr.getLeft()!=null){
            curr=curr.getLeft();
        }
        return curr;
    }
    public boolean isEmpty(){
        return this.nodeCount==0;
    }
    public int size(){
        return this.nodeCount;
    }
    
    
}

File: Python/bst.py
class Node:
    def __init__(self, n=None):
        self.val=n
        self.left=None
        self.right=None
    def getVal(self):
        return self.val
    def setVal(self, n):
        self.val=n
    def getRight(self):
        return self.right
    def getLeft(self):
        return self.left
    def setLeft(self, n=None):
        if(isinstance(n, int)):
            self.left=Node(n)
        else:
            self.left=n
    def setRight(self, n=None):
        if(isinstance(n, int)):
            self.right=Node(n)
        else:
            self.right=n
class BST:
    def __init__(self, n=None):
        if(isinstance(n, int)):
            self.root=Node(n)
        else:
            self.root=n
    def preorderPrint(self):
        self.preorder(self.getRoot())
    def preorder(self, start):
        if start: #not none
            print(start.getVal())
        if(start.getLeft()):
            self.preorder(start.getLeft())
        if(start.getRight()):
            self.preorder(start.getRight())
    def postrderPrint(self):
        self.postorder(self.getRoot())
    def postorder(self, start):
        if start: #not none
            print(start.getVal())
        if(start.getLeft()):
            self.postorder(start.getLeft())
        if(start.getRight()):
            self.postorder(start.getRight())
    def __str__(self):
        #level order traversal
        s=""
        q=[self.root]
        while(q):
            s+=str(q[-1].getVal())+" "
            node=q.pop()
            if(node.getLeft()!=None):
                q=[node.getLeft()]+q
            if(node.getRight()!=None):
                q=[node.getRight()]+q
        return s
    def getRoot(self):
        return self.root
    def size(self):
        q=[self.root]
        count=0
        while(q):
            node=q.pop()
            count+=1
            if(node.getLeft()!=None):
                q=[node.getLeft()]+q
            if(node.getRight()!=None):
                q=[node.getRight()]+q
        return count
    def addNode(self, n):
        head=self.root
        if(self.root==None):
            self.root.setVal(n)
        else:
            curr=self.root
            parent=curr
            while(True):
                parent=curr
                if(n<curr.getVal()):
                    curr=curr.getLeft()
                    if(curr==None):
                        parent.setLeft(n)
                        return head
                elif(n==curr.getVal()):
                    return head
                else:
                    curr=curr.getRight()
                    if(curr==None):
                        parent.setRight(n)
                        return head
    def removeNode(self, n):
        head=self
        curr=self.getRoot()
        parent=curr
        #1. find element to remove if it exists
        #2. replace the node we are removing with its successor if one exists
        if(self.contains(n)):
            if(curr.getVal()==n):
                curr=curr.getLeft()
            return self.remove(curr, parent, head, n, '')
    def remove(self, curr, parent, head, n, dir):
        if(n==curr.getVal()):
            #1. node to remove has no children
            if(curr.getRight()==None and curr.getLeft()==None):
                if(dir=='l'):
                    parent.setLeft(None)
                if(dir=='r'):
                    parent.setRight(None)
                return head
            #2. node to remove has only right
            elif(curr.getRight()!=None and curr.getLeft()==None):
                if(dir=='l'):
                    parent.setLeft(curr.getRight())
                if(dir=='r'):
                    parent.setRight(curr.getRight())
                return head
            #3. node to remove has only left
            elif(curr.getRight()==None and curr.getLeft()!=None):
                if(dir=='l'):
                    parent.setLeft(curr.getLeft())
                if(dir=='r'):
                    parent.setRight(curr.getLeft())
                return head
            #4. node to remove has both
            else:
                if(dir=='l'):
                    parent.setLeft(None)
                if(dir=='r'):
                    parent.setRight(None)
                return head
        elif(n<curr.getVal()):
            return self.remove(curr.getLeft(), curr, head, n,'l')
        elif(n>curr.getVal()):
            return self.remove(curr.getRight(), curr, head, n, 'r')
    def contains(self, n):
        return self.find(self.getRoot(), n)
    def find(self, node, n):
        curr=node
        #4 possibilites:
        #we hit a null node
        if(curr==None):
            print("Value doesn't exist!")
            return False
        #we found it 
        if(n==curr.getVal()):
            return True
        elif(n<curr.getVal()):
            return self.find(curr.getLeft(), n)
        elif(n>curr.getVal()):
            return self.find(curr.getRight(), n)
def isValid(tree):
    return isValidTree(tree.getRoot(), None, '')
def isValidTree(curr, parent, side):
    #go left and right until the end 
    #depth first search until 
    if(curr==None):
        return True
    if(parent!=None):
        if(parent.getVal()>curr.getVal() and side=='l'):
            return False
        if(parent.getVal()<curr.getVal() and side=='r'):
            return False
    x=isValidTree(curr.getRight(), curr, 'l')
    y=isValidTree(curr.getLeft(), curr, 'r')
    return x and y
#1)Find the first common parent of two nodes in the a tree.    
if __name__=='__main__':
    tree=BST(7)
    tree.addNode(20)
    tree.addNode(5)
    tree.addNode(15)
    tree.addNode(10)
    tree=tree.removeNode(7)
    #assert tree.contains(10)==True
    print(tree)
    assert isValid(tree)==True